File: solutions/csharp/log-levels/1/LogLevels.cs
using System.Text.RegularExpressions;

static class LogLine
{
    public static string Message(string logLine)
    {
        return Regex.Match(logLine, @"\[\S+\]: +(.*)").Groups[1].Value.Trim();
    }

    public static string LogLevel(string logLine)
    {
        return Regex.Match(logLine, @"\[(\S+)\]").Groups[1].Value.ToLower();
    }

    public static string Reformat(string logLine)
    {
        var m = Regex.Match(logLine, @"\[(\S+)\]: +(.*)");
        return $"{m.Groups[2].Value.Trim()} ({m.Groups[1].Value.ToLower()})";
    }
}

File: solutions/cpp/log-levels/1/log_levels.cpp
#include <sstream>
#include <string>

namespace log_line {
    std::string message(std::string line) {
        auto i = line.find(": ");
        if(i == std::string::npos) {
            return "";
        }
        return line.substr(i + 2);
    }

    std::string log_level(std::string line) {
        auto a = line.find("[");
        if(a == std::string::npos) {
            return "";
        }
        a += 1;
        auto b = line.find("]");
        if(b == std::string::npos) {
            return "";
        }
        return line.substr(a, b - a);
    }

    std::string reformat(std::string line) {
        auto msg = message(line);
        auto lvl = log_level(line);

        std::stringstream ss;
        ss << msg << " (" << lvl << ")";
        return ss.str();
    }
}  // namespace log_line
